### PR TITLE
Use Clang wrappers and reduce thread spawning and separate out SampledSignals

### PR DESCRIPTION
**Pull request by bramtayl**
This is near total rewrite of the existing code. However, the front end interface remains the same so there should be no breaking changes. The three main changes are:

1) Use clang wrappers (described in the new readme in the gen folder)
2) Reduce thread spawning: instead of spawning a new thread for each read and write, have a single thread for reads and a single thread for writes
3) Add an internal "scribe" interface that allows users to opt out of SampledSignals if they want to access the buffers directly

### CODE REVIEW

File: src/PortAudio.jl [hunk 1]
```diff
@@ -36,43 +35,46 @@ function versioninfo(io::IO = stdout)
     println(io, "Version: ", Pa_GetVersion())
 end
 
+struct Bounds
```
Comment (bramtayl) [inline src/PortAudio.jl:38]: Use to factor the PortAudioDevice fields

File: src/PortAudio.jl [hunk 2]
```diff
@@ -36,43 +35,46 @@ function versioninfo(io::IO = stdout)
...
     )
 end
 
+name(device::PortAudioDevice) = device.name
+
 function devices()
-    ndevices = Pa_GetDeviceCount()
-    infos = PaDeviceInfo[Pa_GetDeviceInfo(i) for i in 0:(ndevices - 1)]
-    PortAudioDevice[PortAudioDevice(info, idx - 1) for (idx, info) in enumerate(infos)]
+    [PortAudioDevice(Pa_GetDeviceInfo(i), i) for i in 0:(Pa_GetDeviceCount() - 1)]
```
Comment (bramtayl) [inline src/PortAudio.jl:75]: avoid allocations

File: src/PortAudio.jl [hunk 3]
```diff
@@ -36,43 +35,46 @@ function versioninfo(io::IO = stdout)
...
+        ),
+        Bounds(
+            info.max_output_channels,
+            info.default_low_output_latency,
+            info.default_high_output_latency,
+        ),
     )
 end
 
+name(device::PortAudioDevice) = device.name
```
Comment (bramtayl) [inline src/PortAudio.jl:72]: define this function consistently where possible

File: src/PortAudio.jl [hunk 4]
```diff
@@ -91,83 +93,50 @@ struct PortAudioStream{T}
...
-        Pa_StartStream(pointer_ref[])
-        this = new(
-            sr, 
-            latency, 
-            pointer_ref, 
-            warn_xruns,
-            recover_xruns,
-            sink_buffer,
-            source_buffer
+function make_parameters(device, channels, T, latency, host_api_specific_stream_info)
```
Comment (bramtayl) [inline src/PortAudio.jl:98]: avoid code duplication

File: src/PortAudio.jl [hunk 5]
```diff
@@ -91,83 +93,50 @@ struct PortAudioStream{T}
...
+    if channels == 0
+        Ptr{Pa_StreamParameters}(0)
+    else
+        Ref(
+            Pa_StreamParameters(
+                device.idx,
+                channels,
+                type_to_fmt[T],
+                latency,
+                convert_nothing(host_api_specific_stream_info),
```
Comment (bramtayl) [inline src/PortAudio.jl:108]: expose all arguments

File: src/PortAudio.jl [hunk 6]
```diff
@@ -91,83 +93,50 @@ struct PortAudioStream{T}
...
+            ),
         )
-        # pre-fill the output stream so we're less likely to underrun
-        prefill_output(this.sink)
+    end
+end
 
-        this
+function fill_max_channels(channels, bounds)
+    if channels === max
```
Comment (bramtayl) [inline src/PortAudio.jl:115]: avoid code duplication. also node that i've replaced the undocuented -1 sentinel and instead you can pass the max function itself. this seems clearer to me, if we like the change i can document it

File: src/PortAudio.jl [hunk 7]
```diff
@@ -91,83 +93,50 @@ struct PortAudioStream{T}
...
-        prefill_output(stream.sink)
+        discard_input(source)
+        prefill_output(sink)
     end
 end
 
-function defaultlatency(devices...)
-    maximum(d -> max(d.highoutputlatency, d.highinputlatency), devices)
+function defaultlatency(input_device, output_device)
+    max(input_device.input_bounds.high_latency, output_device.output_bounds.high_latency)
```
Comment (bramtayl) [inline src/PortAudio.jl:139]: this is a slightly smarter way to go about things. output bounds shouldn't matter for the input device, and vice versa.

File: src/PortAudio.jl [hunk 8]
```diff
@@ -218,29 +187,62 @@ function PortAudioStream(
...
         recover_xruns,
+        Buffer{eltype}(outdev, outchans),
+        Buffer{eltype}(indev, inchans),
     )
+    # pre-fill the output stream so we're less likely to underrun
+    prefill_output(this.sink)
+    this
+end
+
+function device_by_name(device_name)
```
Comment (bramtayl) [inline src/PortAudio.jl:232]: create the error message while we are looping through the devices so we don't have to do it it twice

File: src/PortAudio.jl [hunk 9]
```diff
@@ -303,18 +295,20 @@ function PortAudioStream(fn::Function, args...; kwargs...)
...
+        pointer_ref[] = C_NULL
     end
     nothing
 end
 
 isopen(stream::PortAudioStream) = stream.pointer_ref[] != C_NULL
 
 SampledSignals.samplerate(stream::PortAudioStream) = stream.samplerate
-eltype(stream::PortAudioStream{T}) where {T} = T
+eltype(::Type{PortAudioStream{T}}) where {T} = T
```
Comment (bramtayl) [inline src/PortAudio.jl:311]: eltype should be defined on types, not values

**Comment by bramtayl** [comment]
Ok, so, I figured as long as I'm doing a rewrite for #63, might as well implement it as well. This could probably use an extra set of eyes.

**Comment by bramtayl** [comment]
Note: I'm guessing this will tank coverage because of all the new automatically generated and unused wrappers

**Comment by rob-luke** [comment]
Looking good @bramtayl. Just a few requests.

* Your PRs are quite comprehensive and touch many aspects of the project. As such, can you tag a new version before and after each PR? I think there is no harm is tagging lots of versions, and would let us easily roll back if an issue arrises.
* Could you described in the PRs not just what you are doing, but why. It only needs to be a single line of text. But for this example, what is clang? why would we switch to it? what's the benifit? what's the negative? and most importantly for me as a reviewer, what's the potential complications this issue may raise that I should look for (e.g. is it likely to complicate things in different operating systems? is it likely to add latency to the audio? is it likely to add clicks at audio onset etc).

I appreciate all the love you've been giving this package recently. Awesome stuff.

**Comment by bramtayl** [comment]
> Can you tag a new version before and after each PR?

Sure! I'll tag before merging, if we decide to merge

> What is clang? Why would we switch to it? What's the benefit? What's the negative?

The clang generators will automatically generate wrappers for a C library based on its headers. So everything you see in libportaudio.jl is automatically generated from the package. We would presumably switch if the benefits outweigh the costs, namely:

Benefits:

- If a newer version of portaudio adds more features, we won't have to add new wrappers: clang will handle it for us
- It will be a lot easier to use currently unused features: the wrappers have already been written for us

Costs:

- Even though it does an admirable job, clang doesn't do everything that the original wrappers did, namely, handle errors and set locks. Fortunately, it's very easy to add secondary wrappers, or just do it at point of use. See `get_default_input_device` in this PR for an example.

> What's the potential complications this issue may raise that I should look for

This is entirely a rearrangement of the existing code, so there shouldn't be any performance or functionality changes at all. If there is, it would be due to a mistake on my part.

**Comment by bramtayl** [comment]
So looking at the code, I realized that there were new threads being spawned during each buffer read and write. This takes time, instead, we can just maintain two demons, one for reading, one for writing, and communicate with them via channels. Since I'm making so many other changes, I've gone ahead and done that too. This seems to have totally eliminated xruns on my end, so I've set xrun warnings to true by default.

**Comment by bramtayl** [comment]
I've also tried to pepper in many more comments and make the code more self documenting to aid in review

**Comment by rob-luke** [comment]
Lots of detailed work here 👏 want me to take a listen yet? Or you're still going?

**Comment by bramtayl** [comment]
Still going...sorry I'll label it as WIP. I think the last commit might have reduced performance, and I could simply it a bit too.

**Comment by bramtayl** [comment]
@rob-luke ok I think I'm ready now!

**Comment by rob-luke** [comment]
Hi @bramtayl I pulled your branch to run a few tests and have been receiving overflow errors. I do not get these overflow errors on the master branch.

Here is my terminal demonstrating the overflow warnings:
image


And here is the terminal demonstrating no warnings on master:
image

Is this expected? It may be that you have changed the warning system in this PR, I haven't read the code yet. But you don't describe this change above so I expect its undesired. It may also be that I am installing the new version wrong?

This is being run on MacOS with an external usb microphone. Once we understand this warning then I can test the audio output using my RME FireFace sound card.

File: gen/generator.jl [hunk 10]
```diff
@@ -0,0 +1,16 @@
+using Clang.Generators
```
Comment (rob-luke) [inline gen/generator.jl:1]: Can you add a readme to this directory explaining what it is for?

File: test/runtests.jl [hunk 11]
```diff
@@ -36,151 +66,192 @@ end
...
+            @test Pa_GetStreamCpuLoad(pointer_to) >= 0
+            @test PaErrorCode(handle_status(Pa_AbortStream(pointer_to))) == paNoError
+            @test PaErrorCode(
+                handle_status(Pa_SetStreamFinishedCallback(pointer_to, C_NULL)),
+            ) == paNoError
         end
     end
 end
+
+# sample rates disagree
```
Comment (rob-luke) [inline test/runtests.jl:257]:
```suggestion
```

**Review by rob-luke — CHANGES_REQUESTED** [review]

**Comment by bramtayl** [comment]
Yes, I changed `warn_xruns` to `true` by default as described here: https://github.com/JuliaAudio/PortAudio.jl/pull/79#issuecomment-866295761. The xrun warnings are disappointing though; I had hoped they might have disappeared. What results do you get if you add `warn_xruns = true` on the master branch? Do you get the same number of xruns?

File: gen/generator.jl (hunk 10 above)
Comment (bramtayl) [reply gen/generator.jl:1]: Ok, I added a README, hopefully it makes sense!

**Comment by rob-luke** [comment]
Thanks for clarifying @bramtayl. Could you summarise all the changes (specifically ones affecting end users like the warnings that I observed) at the top of the pull request in the initial comment. This makes it easier to review, whereas comments scattered throughout commits get easily lost.

Summarising the changes at the top also allows me to more easily go through and check that all the changes work when I test them locally. As it stands now, its hard for me to know what to test.

> What results do you get if you add warn_xruns = true on the master branch? Do you get the same number of xruns?

Using the USB microphone with master I get one overflow. With this branch I get two overflows.

Using my RME I get to overflows on each branch.

I dont use audio input very often, so I cant comment if this is expected or not, but I imagine its not good.

**Comment by bramtayl** [comment]
Hmm, could you try again, but run the full test suite, adding `warn_xruns = true` to each call to PortAudioStream on master? I get 5 xruns on master and 1 on this branch.

**Comment by bramtayl** [comment]
I tried my hand at making performance improvements; maybe it helped a little? Also, I edited the initial comment to summarize the changes here.

**Comment by bramtayl** [comment]
Ok, no more xruns on my end

**Comment by rob-luke** [comment]
Wow! Ok I'll run the test suite and send you my results today.

**Comment by rob-luke** [comment]
Running this branch I still get a few over/underflows. I have pasted the output below for you incase that helps.

For such a major rewrite as this it would be good to get some input and testing from others. Once you are happy with this PR lets ping a few people for a review. 

<details>
  <summary>PR output</summary>

```

➜  PortAudio git:(bramtayl/master) ✗ julia --project=. test/runtests.jl
9-element Vector{PortAudioDevice}:
 "Built-in Input" 2→0
 "Built-in Line Input" 2→0
 "Built-in Output" 0→2
 "Built-in Line Output" 0→2
 "Built-in Line Output" 0→2
 "Built-in Digital Output" 0→2
 "Unknown USB Audio Device" 1→0
 "Fireface UCX (23634778)" 18→18
 "ZoomAudioDevice" 2→2
Test Summary:       | Pass  Total
Tests without sound |   16     16
Recording...
┌ Warning: libportaudio: Input overflowed
└ @ PortAudio ~/.julia/dev/PortAudio/src/PortAudio.jl:100
┌ Warning: libportaudio: Input overflowed
└ @ PortAudio ~/.julia/dev/PortAudio/src/PortAudio.jl:100
Playing back recording...
Testing pass-through
Interactive tests: Test Failed at /Users/rluke/.julia/dev/PortAudio/test/runtests.jl:144
  Expression: sprint(show, sink) == "2 channel sink: $(repr(default_input_device_name))"
   Evaluated: "2 channel sink: \"Fireface UCX (23634778)\"" == "2 channel sink: \"Built-in Input\""
Stacktrace:
 [1] macro expansion
   @ ~/.julia/dev/PortAudio/test/runtests.jl:144 [inlined]
 [2] macro expansion
   @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
 [3] macro expansion
   @ ~/.julia/dev/PortAudio/test/runtests.jl:123 [inlined]
 [4] macro expansion
   @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
 [5] top-level scope
   @ ~/.julia/dev/PortAudio/test/runtests.jl:122
Interactive tests: Test Failed at /Users/rluke/.julia/dev/PortAudio/test/runtests.jl:145
  Expression: sprint(show, source) == "2 channel source: $(repr(default_output_device_name))"
   Evaluated: "2 channel source: \"Built-in Input\"" == "2 channel source: \"Fireface UCX (23634778)\""
Stacktrace:
 [1] macro expansion
   @ ~/.julia/dev/PortAudio/test/runtests.jl:145 [inlined]
 [2] macro expansion
   @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
 [3] macro expansion
   @ ~/.julia/dev/PortAudio/test/runtests.jl:123 [inlined]
 [4] macro expansion
   @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
 [5] top-level scope
   @ ~/.julia/dev/PortAudio/test/runtests.jl:122
┌ Warning: libportaudio: Input overflowed
└ @ PortAudio ~/.julia/dev/PortAudio/src/PortAudio.jl:100
┌ Warning: libportaudio: Output underflowed
└ @ PortAudio ~/.julia/dev/PortAudio/src/PortAudio.jl:100
done
┌ Warning: libportaudio: Input overflowed
└ @ PortAudio ~/.julia/dev/PortAudio/src/PortAudio.jl:100
Constructors: Error During Test at /Users/rluke/.julia/dev/PortAudio/test/runtests.jl:203
  Got exception outside of a @test
  DomainError with 2:
  2 exceeds max output channels for Built-in Input
  Stacktrace:
    [1] fill_max_channels
      @ ~/.julia/dev/PortAudio/src/PortAudio.jl:546 [inlined]
    [2] PortAudioStream(input_device::PortAudioDevice, output_device::PortAudioDevice, input_channels::Int64, output_channels::Int64; Sample::Type, samplerate::Nothing, latency::Nothing, warn_xruns::Bool, frames_per_buffer::Int64, flags::UInt64, call_back::Ptr{Nothing}, user_data::Ptr{Nothing}, input_info::Ptr{Nothing}, output_info::Ptr{Nothing}, stream_lock::ReentrantLock, writer::PortAudio.SampledSignalsWriter{Float32}, reader::PortAudio.SampledSignalsReader{Float32})
      @ PortAudio ~/.julia/dev/PortAudio/src/PortAudio.jl:628
    [3] PortAudioStream
 t/runtests.jl+                                                                                                                                                                      buffers
      @ ~/.julia/dev/PortAudio/src/PortAudio.jl:625 [inlined]
    [4] #PortAudioStream#17
      @ ~/.julia/dev/PortAudio/src/PortAudio.jl:742 [inlined]
    [5] PortAudioStream
      @ ~/.julia/dev/PortAudio/src/PortAudio.jl:742 [inlined]
    [6] #PortAudioStream#18
      @ ~/.julia/dev/PortAudio/src/PortAudio.jl:757 [inlined]
    [7] PortAudioStream (repeats 2 times)
      @ ~/.julia/dev/PortAudio/src/PortAudio.jl:757 [inlined]
    [8] PortAudioStream(do_function::var"#14#25", arguments::String; keywords::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
      @ PortAudio ~/.julia/dev/PortAudio/src/PortAudio.jl:775
    [9] PortAudioStream(do_function::Function, arguments::String)
      @ PortAudio ~/.julia/dev/PortAudio/src/PortAudio.jl:775
   [10] macro expansion
      @ ~/.julia/dev/PortAudio/test/runtests.jl:207 [inlined]
   [11] macro expansion
      @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
   [12] macro expansion
      @ ~/.julia/dev/PortAudio/test/runtests.jl:204 [inlined]
   [13] macro expansion
      @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
   [14] top-level scope
      @ ~/.julia/dev/PortAudio/test/runtests.jl:122
   [15] include(mod::Module, _path::String)
      @ Base ./Base.jl:386
   [16] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:285
   [17] _start()
      @ Base ./client.jl:485
Test Summary:                   | Pass  Fail  Error  Total
Tests with sound                |   25     2      1     28
  Interactive tests             |    7     2             9
  Samplerate-converting writing |                    No tests
  Queued Writing                |    2                   2
  Queued Reading                |    2                   2
  Constructors                  |    1            1      2
  Errors with sound             |    3                   3
  libportaudio with sound       |   10                  10
```
</details>



<details>
  <summary>master output</summary>
  
```

➜  PortAudio git:(master) ✗ julia --project=. test/runtests.jl
┌ Debug:
└ @ PortAudio ~/.julia/dev/PortAudio/src/PortAudio.jl:23
Test Summary:  | Pass  Total
Debug messages |    2      2
Test Summary:   | Pass  Total
PortAudio Tests |    3      3
┌ Debug:
└ @ Main ~/.julia/dev/PortAudio/src/PortAudio.jl:23
Recording...
┌ Debug:
└ @ PortAudio ~/.julia/dev/PortAudio/src/PortAudio.jl:23
Playing back recording...
┌ Debug:
└ @ PortAudio ~/.julia/dev/PortAudio/src/PortAudio.jl:23
Testing pass-through
┌ Debug:
└ @ PortAudio ~/.julia/dev/PortAudio/src/PortAudio.jl:23
Open Default Device: Test Failed at /Users/rluke/.julia/dev/PortAudio/test/runtests.jl:78
  Expression: sprint(show, sink) == "2-channel PortAudioSink{Float32}($(repr(default_indev)))"
   Evaluated: "2-channel PortAudioSink{Float32}(\"Fireface UCX (23634778)\")" == "2-channel PortAudioSink{Float32}(\"Built-in Input\")"
Stacktrace:
 [1] macro expansion
   @ ~/.julia/dev/PortAudio/test/runtests.jl:78 [inlined]
 [2] macro expansion
   @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
 [3] macro expansion
   @ ~/.julia/dev/PortAudio/test/runtests.jl:62 [inlined]
 [4] macro expansion
   @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
 [5] top-level scope
   @ ~/.julia/dev/PortAudio/test/runtests.jl:61
Open Default Device: Test Failed at /Users/rluke/.julia/dev/PortAudio/test/runtests.jl:80
  Expression: sprint(show, source) == "2-channel PortAudioSource{Float32}($(repr(default_outdev)))"
   Evaluated: "2-channel PortAudioSource{Float32}(\"Built-in Input\")" == "2-channel PortAudioSource{Float32}(\"Fireface UCX (23634778)\")"
Stacktrace:
 [1] macro expansion
   @ ~/.julia/dev/PortAudio/test/runtests.jl:80 [inlined]
 [2] macro expansion
   @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
 [3] macro expansion
   @ ~/.julia/dev/PortAudio/test/runtests.jl:62 [inlined]
 [4] macro expansion
   @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
 [5] top-level scope
   @ ~/.julia/dev/PortAudio/test/runtests.jl:61
done
┌ Debug:
└ @ PortAudio ~/.julia/dev/PortAudio/src/PortAudio.jl:23
┌ Debug:
└ @ PortAudio ~/.julia/dev/PortAudio/src/PortAudio.jl:23
┌ Debug:
└ @ PortAudio ~/.julia/dev/PortAudio/src/PortAudio.jl:23
┌ Debug:
└ @ PortAudio ~/.julia/dev/PortAudio/src/PortAudio.jl:23
Test Summary:                   | Pass  Fail  Total
Local Tests                     |   20     2     22
  Open Default Device           |    8     2     10
  Samplerate-converting writing |             No tests
  Open Device by name           |    2            2
  Error handling                |    6            6
  Queued Writing                |    2            2
  Queued Reading                |    2            2
ERROR: LoadError: Some tests did not pass: 20 passed, 2 failed, 0 errored, 0 broken.
in expression starting at /Users/rluke/.julia/dev/PortAudio/test/runtests.jl:49

```

</details>

**Comment by bramtayl** [comment]
It's also interesting that all four of your errors are from the input. When googling fireface I get "the Fireface UC provides revolutionary ultra-low latencies". Maybe the xrun errors would go away if you bumped up the latency a bit?

**Comment by bramtayl** [comment]
Ok, the show errors might have been caused by a source/sink issue I noticed and just fixed. But there's still something funny going on. Is it possible that the devices changed mid-test?

**Comment by rob-luke** [comment]
> did you add warn_xruns=true to the PortAudioStream calls?

No 😢, but now I did. See the output below.

<details>
  <summary>master output with xrun warnings</summary>
  
```

➜  PortAudio git:(masterwxrun) ✗ julia --project=. test/runtests.jl
Test Summary:  | Pass  Total
Debug messages |    2      2
Test Summary:   | Pass  Total
PortAudio Tests |    3      3
Recording...
┌ Warning: libportaudio: Input overflowed
└ @ PortAudio ~/.julia/dev/PortAudio/src/libportaudio.jl:329
┌ Warning: libportaudio: Input overflowed
└ @ PortAudio ~/.julia/dev/PortAudio/src/libportaudio.jl:329
Playing back recording...
Testing pass-through
Open Default Device: Test Failed at /Users/rluke/.julia/dev/PortAudio/test/runtests.jl:77
  Expression: sprint(show, sink) == "2-channel PortAudioSink{Float32}($(repr(default_indev)))"
   Evaluated: "2-channel PortAudioSink{Float32}(\"Fireface UCX (23634778)\")" == "2-channel PortAudioSink{Float32}(\"Built-in Input\")"
Stacktrace:
 [1] macro expansion
   @ ~/.julia/dev/PortAudio/test/runtests.jl:77 [inlined]
 [2] macro expansion
   @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
 [3] macro expansion
   @ ~/.julia/dev/PortAudio/test/runtests.jl:61 [inlined]
 [4] macro expansion
   @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
 [5] top-level scope
   @ ~/.julia/dev/PortAudio/test/runtests.jl:60
Open Default Device: Test Failed at /Users/rluke/.julia/dev/PortAudio/test/runtests.jl:79
  Expression: sprint(show, source) == "2-channel PortAudioSource{Float32}($(repr(default_outdev)))"
   Evaluated: "2-channel PortAudioSource{Float32}(\"Built-in Input\")" == "2-channel PortAudioSource{Float32}(\"Fireface UCX (23634778)\")"
Stacktrace:
 [1] macro expansion
   @ ~/.julia/dev/PortAudio/test/runtests.jl:79 [inlined]
 [2] macro expansion
   @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
 [3] macro expansion
   @ ~/.julia/dev/PortAudio/test/runtests.jl:61 [inlined]
 [4] macro expansion
   @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
 [5] top-level scope
   @ ~/.julia/dev/PortAudio/test/runtests.jl:60
┌ Warning: libportaudio: Input overflowed
└ @ PortAudio ~/.julia/dev/PortAudio/src/libportaudio.jl:329
done
Test Summary:                   | Pass  Fail  Total
Local Tests                     |   20     2     22
  Open Default Device           |    8     2     10
  Samplerate-converting writing |             No tests
  Open Device by name           |    2            2
  Error handling                |    6            6
  Queued Writing                |    2            2
  Queued Reading                |    2            2
ERROR: LoadError: Some tests did not pass: 20 passed, 2 failed, 0 errored, 0 broken.
in expression starting at /Users/rluke/.julia/dev/PortAudio/test/runtests.jl:48
```

</details>


> Is it possible that the devices changed mid-test?

I am quite sure it didn't, unless the test code makes it. I have run the tests multiple times and always get the same output.

> When googling fireface

I think the tests are actually using the `"Built-in Input" 2→0` which is some motherboard device ive never used. So I dont think the test are using the Fireface. I would prefer the tests do use the Fireface as its a very standard piece of audio equipment in labs. Is there a way I can force the tests to use that device? I tried setting the `input_index` variable so that the Fireface was used, but then I got stacks of new errors.

> and just fixed.

Yes you did!! And that works great. Here is the new output of the test on my pc incase that helps. I am still getting the constructor error. 


<details>
  <summary>PR</summary>
  
```
➜  PortAudio git:(bramtayl/master) ✗ julia --project=. test/runtests.jl
8-element Vector{PortAudioDevice}:
 "Built-in Input" 2→0
 "Built-in Line Input" 2→0
 "Built-in Output" 0→2
 "Built-in Line Output" 0→2
 "Built-in Line Output" 0→2
 "Built-in Digital Output" 0→2
 "Fireface UCX (23634778)" 18→18
 "ZoomAudioDevice" 2→2
Test Summary:       | Pass  Total
Tests without sound |   16     16
Recording...
┌ Warning: libportaudio: Input overflowed
└ @ PortAudio ~/.julia/dev/PortAudio/src/PortAudio.jl:100
┌ Warning: libportaudio: Input overflowed
└ @ PortAudio ~/.julia/dev/PortAudio/src/PortAudio.jl:100
Playing back recording...
Testing pass-through
done
┌ Warning: libportaudio: Input overflowed
└ @ PortAudio ~/.julia/dev/PortAudio/src/PortAudio.jl:100
Constructors: Error During Test at /Users/rluke/.julia/dev/PortAudio/test/runtests.jl:203
  Got exception outside of a @test
  DomainError with 2:
  2 exceeds max output channels for Built-in Input
  Stacktrace:
    [1] fill_max_channels
      @ ~/.julia/dev/PortAudio/src/PortAudio.jl:546 [inlined]
    [2] PortAudioStream(input_device::PortAudioDevice, output_device::PortAudioDevice, input_channels::Int64, output_channels::Int64; Sample::Type, samplerate::Nothing, latency::Nothing, warn_xruns::Bool, frames_per_buffer::Int64, flags::UInt64, call_back::Ptr{Nothing}, user_data::Ptr{Nothing}, input_info::Ptr{Nothing}, output_info::Ptr{Nothing}, stream_lock::ReentrantLock, writer::PortAudio.SampledSignalsWriter{Float32}, reader::PortAudio.SampledSignalsReader{Float32})
      @ PortAudio ~/.julia/dev/PortAudio/src/PortAudio.jl:628
    [3] PortAudioStream
      @ ~/.julia/dev/PortAudio/src/PortAudio.jl:625 [inlined]
    [4] #PortAudioStream#17
      @ ~/.julia/dev/PortAudio/src/PortAudio.jl:742 [inlined]
    [5] PortAudioStream
      @ ~/.julia/dev/PortAudio/src/PortAudio.jl:742 [inlined]
    [6] #PortAudioStream#18
      @ ~/.julia/dev/PortAudio/src/PortAudio.jl:757 [inlined]
    [7] PortAudioStream (repeats 2 times)
      @ ~/.julia/dev/PortAudio/src/PortAudio.jl:757 [inlined]
    [8] PortAudioStream(do_function::var"#14#25", arguments::String; keywords::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
      @ PortAudio ~/.julia/dev/PortAudio/src/PortAudio.jl:775
    [9] PortAudioStream(do_function::Function, arguments::String)
      @ PortAudio ~/.julia/dev/PortAudio/src/PortAudio.jl:775
   [10] macro expansion
      @ ~/.julia/dev/PortAudio/test/runtests.jl:207 [inlined]
   [11] macro expansion
      @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
   [12] macro expansion
      @ ~/.julia/dev/PortAudio/test/runtests.jl:204 [inlined]
   [13] macro expansion
      @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
   [14] top-level scope
      @ ~/.julia/dev/PortAudio/test/runtests.jl:122
   [15] include(mod::Module, _path::String)
      @ Base ./Base.jl:386
   [16] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:285
   [17] _start()
      @ Base ./client.jl:485
Test Summary:                   | Pass  Error  Total
Tests with sound                |   27      1     28
  Interactive tests             |    9             9
  Samplerate-converting writing |              No tests
  Queued Writing                |    2             2
  Queued Reading                |    2             2
  Constructors                  |    1      1      2
  Errors with sound             |    3             3
  libportaudio with sound       |   10            10

```

</details>

**Comment by bramtayl** [comment]
Am I right in counting three xruns each on master and the PR? Not great but not terrible then.

> Is there a way I can force the tests to use that device? I tried setting the input_index variable so that the Fireface was used, but then I got stacks of new errors.

The tests should be using the default input and output devices as given by PortAudio. I don't know where PortAudio gets that information from? I could add a line at the start like `testing_input_device_index = [default]` so you could edit the script more easily.

> I tried setting the `input_index` variable so that the Fireface was used, but then I got stacks of new errors.

What errors?

**Comment by bramtayl** [comment]
With the domain error, maybe we could add an `adjust = false` keyword that if enabled would set the requested device count to the max if the max is exceeded? Then we could enable it in all the tests?

**Comment by bramtayl** [comment]
Ok, I added an `adjust_channels = false` to enable the tests not to exceed the max, and pulled out the device names to make them easier to hardcode into the tests

**Comment by bramtayl** [comment]
Updates: I added real docs (both to document new features, and also cause we didn't really have them before). The doctests don't seem to run, I think it has something to do with https://github.com/JuliaIO/Suppressor.jl/issues/28

**Comment by bramtayl** [comment]
Well, darn, didn't mean to merge this but guess it's too late now...I do think it should be pretty good quality code though, I've been working on it long enough. I don't think there's anything we can do about the doctest issue; I'm guessing the problem is that redirect_stderr and documenter are competing for access to stderr

**Comment by rob-luke** [comment]
Please use squash and merge when merging. The commit history is now becoming littered with junk like "fix".

@bramtayl  Could you try and look up some git magic to retrospectively fix your commit history from this PR?  I am sure it can be done.

**Comment by bramtayl** [comment]
Ok, sorry. I've squashed all the commits and forced pushed, so hopefully that should be better?

**Comment by bramtayl** [comment]
Hey, and look at the new docs! https://juliaaudio.github.io/PortAudio.jl/dev/

**Comment by bramtayl** [comment]
@JuliaRegistrator register

**Pull request merged**